Tests/UnitTests/Core/TrackFinding/CombinatorialKalmanFilterTests.cpp

```diff
@@ -22,327 +20,181 @@
-    double rPhi = 0_degree + 0.0002 * gauss(generator);
-    double rTheta = 90_degree + 0.0002 * gauss(generator);
-    CurvilinearTrackParameters rStart(makeVector4(rPos, 42_ns), rPhi, rTheta,
-                                      1_GeV, 1_e, cov);
-
-    const Surface* rSurface = &rStart.referenceSurface();
-
-    auto logger =
-        getDefaultLogger("CombinatorialKalmanFilter", Logging::VERBOSE);
-    CombinatorialKalmanFilterOptions<TestSourceLinkCalibrator,
-                                     CKFSourceLinkSelector>
-        ckfOptions(tgContext, mfContext, calContext, TestSourceLinkCalibrator(),
-                   CKFSourceLinkSelector(sourcelinkSelectorConfig),
-                   LoggerWrapper{*logger}, PropagatorPlainOptions(), rSurface);
-
-    // Find the track(s)
-    auto combKalmanFilterRes = cKF.findTracks(sourcelinks, rStart, ckfOptions);
-    BOOST_CHECK(combKalmanFilterRes.ok());
-
-    auto foundTrack = *combKalmanFilterRes;
-    auto& fittedStates = foundTrack.fittedStates;
-    auto& trackTips = foundTrack.trackTips;
-
-    for (const auto& tip : trackTips) {
-      std::vector<size_t> trackIds;
-      fittedStates.visitBackwards(tip, [&](const auto& trackState) {
-        trackIds.push_back(trackState.uncalibrated().sourceId);
-      });
-
-      BOOST_CHECK_EQUAL(trackIds.size(), 6);
-
-      size_t numFakeHit = 0;
-      for (const auto& id : trackIds) {
-        numFakeHit = numFakeHit + (id != trackId ? 1 : 0);
-      }
-
-      // Check if there are fake hits from other tracks
-      BOOST_CHECK_EQUAL(numFakeHit, 0);
+    for (auto& sl : measurements.sourceLinks) {
+      sourceLinks.emplace_back(std::move(sl));
     }
+  }
 
-    // Construct a starting parameter near the tracker exit
-    Vector3D rPosOuter(-1. * rPos.x(), rPos.y(), rPos.z());
-    CurvilinearTrackParameters rStartOuter(makeVector4(rPosOuter, 42_ns), rPhi,
-                                           rTheta, 1_GeV, 1_e, cov);
-    // Reset the navigation direction
-    ckfOptions.propagatorPlainOptions.direction = backward;
-    // Find the track(s)
-    combKalmanFilterRes = cKF.findTracks(sourcelinks, rStartOuter, ckfOptions);
-    BOOST_CHECK(combKalmanFilterRes.ok());
+  // use one of the parameter surfaces as reference surface for the
+  // all reconstructed tracks
+  Acts::CombinatorialKalmanFilterOptions<TestSourceLinkCalibrator,
+                                         Acts::CKFSourceLinkSelector>
+      ckfOptions(geoCtx, magCtx, calCtx, TestSourceLinkCalibrator(),
+                 Acts::CKFSourceLinkSelector(selectBestSourceLinkOnly),
+                 Acts::LoggerWrapper{*ckfLogger},
+                 Acts::PropagatorPlainOptions(),
+                 &tracks.front().referenceSurface());
+
+  // run the CKF for each initial track state
+  for (size_t trackId = 0u; trackId < tracks.size(); ++trackId) {
+    // find the tracks
+    auto res = ckfZero.findTracks(sourceLinks, tracks[trackId], ckfOptions);
```

Comment:
So we use exactly the same starting parameters for the simulation and reconstruction? I guess it should be fine in this test here.

Reply:
Yes, I do.

Let me explain my reasoning here: Since this is a unit test,  I want to test the basic functionality of the algorithm in question and I want to do so reliably and reproducible. Ideally, I use input data for which I know what to expect. This is clear when using the identical start parameters. If I smear them but want to still pick up all hits, I need to smear them only a little bit; at some point I might miss a hit or run into numerical issues, but where that happens is not something that I can predict in advance. Yet, smearing the input parameters only a little bit then does not really test anything different then using the true parameters (since we already have smearing in the measurements).

If I would want to really test the effect of the difference between initial input and true parameters, you would either have to do a systematic scan over all possible displacement values, or do a full statistical analysis with a large set of smeared inputs. Both of which, I consider outside the scope of a unit test.

Reply:
That's indeed fair enough! I agree with you.